driver-core/src/main/com/mongodb/operation/AsyncQueryBatchCursor.java

```diff
@@ -130,41 +120,56 @@ public boolean isClosed() {
         return closed;
     }
 
-    ServerCursor getServerCursor() {
-        return cursor;
+    private boolean getAndSetClosed() {
+        boolean oldClosed;
+        synchronized (lock) {
+            oldClosed = closed;
+            closed = true;
+        }
+        return oldClosed;
+    }
+
+    private ServerCursor getAndSetCursor(final ServerCursor cursor) {
+        ServerCursor oldCursor;
+        synchronized (lock) {
+            oldCursor = this.cursor;
+            this.cursor = cursor;
+        }
+        return oldCursor;
     }
 
     private boolean limitReached() {
         return Math.abs(limit) != 0 && count >= Math.abs(limit);
     }
 
-    private void getMore(final SingleResultCallback<List<T>> callback) {
+    private void getMore(final ServerCursor cursor, final SingleResultCallback<List<T>> callback) {
         connectionSource.getConnection(new SingleResultCallback<AsyncConnection>() {
```

Comment:
I'm still unclear how `connectionSource` can have a reference count of zero.  We retain it in the constructor and it is released via `killCursorOnClose()`.  If the limit is reached, then we call `killCursor()`.  Here the connection source is retained there to ensure `close()` manages the final reference.

However, I now changed it so `next()` sets the cursor to null, now only the getMore can set the cursor.